WordPress/gutenberg
# Post Date: Allow user to pick Site Default, a suggested date format, or a custom date format

## Description

Closes https://github.com/WordPress/gutenberg/issues/38751.

https://user-images.githubusercontent.com/612155/157785443-17bec04b-f7a6-4113-a66c-22d63697dc64.mp4

Changes the Post Date block so that you can choose between using the site's default date format, a suggested format, or a completely custom format.

## How I arrived at the suggested date formats

_Which formats should we suggest?_ is a difficult question as different locales have different date formatting conventions.

In my opinion there are a few requirements:

1. Users should see suggestions that are appropriate for their locale.
2. We shouldn't overwhelm users with too many options.
3. Theme developers should be able to do basically whatever they want.

We can meet requirement 3 by allowing the theme developer to type any arbitrary date format into a text field.

If we pick a small set of suggested formats, as in https://github.com/WordPress/gutenberg/issues/38751#issuecomment-1039113532, then we will meet requirement 2 but struggle to meet requirement 1. This is because any list we come up with is bound to not be appropriate in some locales. For example, 4/20/22 is typical in US English but 20/04/2022 is typical elsewhere. Or, "April 20, 2022" makes sense in US English but this would be "20 April 2022" in UK English and "20. April 2022" in German.

We could flood the dropdown with hundreds of different options so as to cover all locales and meet requirement 1, but then this would be an overwhelming amount of options and therefore not meet requirement 2.

So what I've opted to do is to start from four categories: short, medium, long, and full. This is inspired by [what Apple has in System Preferences](https://user-images.githubusercontent.com/612155/156981568-28918f2e-65b7-4bcb-bea7-e713b9901b5a.png). For the short, medium, and long categories, I added the option of showing the time as well. This gives seven options. Each option is localised depending on the site's language. That is, the suggested formats will change depending on what language WordPress is set to.

| Option | US English `en_US` | UK English `en_GB` | German `de` |
| - | - | - | - |
| Short | 4/20/2022 | 20/04/2022 | 20.04.2022 |
| Medium | Apr 20, 2022 | 20 Apr 2022 | 20. Apr 2022 |
| Long | April 20, 2022 | 20 April 2022 | 20. April 2022 |
| Full | Wednesday, April 20, 2022 | Wednesday, 20 April 2022 | Mittwoch, 20. April 2022 |
| Short + time | 4/20/2022 4:20 AM | 20/04/2022 4:20 am | 20.04.2022 04:20 |
| Medium + time | Apr 20, 2022 4:20 AM | 20 Apr 2022 4:20 am | 20. Apr 2022 04:20 |
| Long + time | April 20, 2022 4:20 AM | 20 April 2022 4:20 am | 20. April 2022 04:20 |
| ISO | 2022-04-20 | 2022-04-20 | 2022-04-20 |

## Notes

- A `null` `attributes.format` is what denotes that the user has selected _Site default_. This is the default.
- Existing Post Date blocks will remain as they are, even if they use the site's default format. I don't think we can properly migrate this as we don't know whether `attributes.format === siteFormat` because the user intended that or if it's just a coincidence.
- We don't need to access the `@wordpress/date` settings as we can get everything we need from `/wp/v2/settings` (the `site` entity).
- ~I need to update the Comment Date block with the same changes. (Maybe the dropdown should be a block editor component? (Or a block supports? 😬))~
- I removed a bunch of unnecessary string/array manipulation from the `is12Hour` calculation and fixed the logic of this. Previously it was only looking for `a` but technically a 12 hour format string could just be e.g. `g:i` with no `a`.

## Testing Instructions

1. Open the site editor.
6. Find a Post Date block and go into its block settings.
7. Experiment with different _Format settings_.
8. If _Site default_ is selected, check that changing the date setting in _Options → General_ changes how the Post Date appears in the frontend.
9. If _Custom_ is selected, check that you can enter a custom date format string.

## Checklist:
